merlion/models/utils/rolling_window_dataset.py

```diff
@@ -0,0 +1,192 @@
+        self.label_axis = label_axis
+
+        self._valid_rolling_steps = len(data) - self.maxlags - self.forecast_steps
+        self._data_len = len(data)
+
+    @property
+    def valid_rolling_steps(self):
+        return self._valid_rolling_steps
+
+    def __len__(self):
+        return self._data_len
+
+    def __iter__(self):
+
+        if self.batch_size is None or self.batch_size < 1:
+            if self.label_axis == 0:
+                yield self._get_entire_train_data_along_time()
+            elif self.label_axis == 1:
+                yield self._get_entire_train_data_along_sequence()
+            return
+
+        if self.batch_size == 1:
+            yield from self._get_iterator()
+        elif self.batch_size > 1:
+            batch = list()
+            for i in self._get_iterator():
+                batch.append(i)
+                if len(batch) >= self.batch_size:
+                    yield batch
+                    batch = list()
+
+    def _get_iterator(self):
+        """
+        For example, if the input is like the following
+
+            for self.label_axis = 0, we roll the window along the time axis
+            .. code-block:: python
+
+                data = (t0, (1,2,3)),
+                       (t1, (4,5,6)),
+                       (t2, (7,8,9)),
+                       (t3, (10,11,12))
+                target_seq_index = 0
+                maxlags = 2
+                forecast_steps = 2
+
+        Then we may yield the following for the first time
+
+            .. code-block:: python
+                (
+                    (t0, (1,2,3)),
+                    (t1, (4,5,6))
+                ),
+                (
+                    (t2, (7))
+                    (t3, (10))
+                )
+
+        """
+        order = np.random.permutation(self.valid_rolling_steps + 1) if self.shuffle \
+            else range(self.valid_rolling_steps + 1)
+        for i in order:
+            j = i + self.maxlags
+            if self.label_axis == 0:
+                if self.ts_index:
+                    past_ts = self.data_ts[i: j]
+                    future_ts = self.target_ts[j: j + self.forecast_steps]
+                    yield past_ts, future_ts
+                else:
+                    past = self.data[i: j]
+                    future = self.target[j: j + self.forecast_steps]
+                    future_timestamp = self.target_timestamp[j: j + self.forecast_steps]
+                    yield past, future, future_timestamp
+            elif self.label_axis == 1:
+                if self.ts_index:
+                    past_ts = self.data_ts[i: j]
+                    future_ts = self.data_ts[j]
+                    yield past_ts, future_ts
+                else:
+                    past = self.data[i: j]
+                    future = self.data[j]
+                    future_timestamp = np.atleast_1d(self.target_timestamp[j])
+                    yield past, future, future_timestamp
```

Comment:
Why not just `yield self[i]`?

Reply:
To me this is more easy to read, which is past or the future

Reply:
sounds good. I worked on __iter__ and finally added an __getitem__, I forgot to merge the two